new-new-packages/compiler/src/compiler/ast.rs

```diff
@@ -88,3 +96,144 @@ impl Deref for AstString {
         &self.value
     }
 }
+
+#[derive(Debug, PartialEq, Eq, Clone, Hash)]
+pub enum AstError {
+    UnexpectedPunctuation,
+    TextWithoutClosingQuote,
+    ParenthesizedWithoutClosingParenthesis,
+    CallOfANonIdentifier,
+    StructWithNonStructField,
+    StructWithoutClosingBrace,
+    ColonMissingAfterStructKey,
+    NonCommaAfterStructValue,
+    ExpectedParameter,
+    LambdaWithoutClosingCurlyBrace,
+}
+
+pub trait CollectErrors {
+    fn collect_errors(self, errors: &mut Vec<CompilerError>);
+}
+impl CollectErrors for Ast {
+    fn collect_errors(self, errors: &mut Vec<CompilerError>) {
+        match self.kind {
+            AstKind::Int(_) => {}
+            AstKind::Text(_) => {}
+            AstKind::Identifier(_) => {}
+            AstKind::Symbol(_) => {}
+            AstKind::Struct(struct_) => {
+                for (key, value) in struct_.fields {
+                    key.collect_errors(errors);
+                    value.collect_errors(errors);
+                }
+            }
+            AstKind::Lambda(lambda) => lambda.body.collect_errors(errors),
+            AstKind::Call(call) => call.arguments.collect_errors(errors),
+            AstKind::Assignment(assignment) => assignment.body.collect_errors(errors),
+            AstKind::Error {
+                child,
+                errors: recovered_errors,
+            } => {
+                errors.append(&mut recovered_errors.clone());
```

Comment:
I think this might be more efficient:

```suggestion
                errors.extend(recovered_errors.into_iter());
```

(Or maybe just `iter`)